File: src/main/java/dev/psygamer/wireframe/event/api/Event.java
package dev.psygamer.wireframe.event.api;

public abstract class Event {
	
	public enum Priority {
		HIGHEST, HIGH, NORMAL, LOW, LOWEST
	}
	
	private boolean canceled = false;
	
	public abstract boolean isCancelable();
	
	public boolean isCanceled() {
		return this.canceled;
	}
	
	public void setCanceled(final boolean canceled) {
		if (!isCancelable()) {
			throw new UnsupportedOperationException(
					"Attempted to call Event#setCanceled() on a non-cancelable event of type: "
							+ this.getClass()
								  .getCanonicalName()
			);
		}
		
		this.canceled = canceled;
	}
}

File: src/main/kotlin/dev/psygamer/wireframe/event/api/Event.kt
package dev.psygamer.wireframe.event.api

abstract class Event {

	var isCanceled = false
		set(canceled) {
			if (!isCancelable) {
				throw UnsupportedOperationException(
					"Attempted to call Event#setCanceled() on a non-cancelable event of type: ${this.javaClass.canonicalName}!"
				)
			}

			field = canceled
		}

	abstract val isCancelable: Boolean
}